### PR TITLE
Add a new TokenAuthentication lib

### PR DESCRIPTION
**Pull request by nZac**

### CODE REVIEW

File: keg_auth/model/__init__.py [hunk 1]
```diff
@@ -274,6 +278,70 @@ def token_generate(self):
...
+        self.token = raw = self.generate_raw_auth_token(**kwargs)
+        return raw
+
+    def verify_token(self, token):
+        if not token or not self.token:
+            return False
+
+        return self.token.context.verify(token, self.token.hash)
+
+    def generate_api_token(self, token=None):
```
[the remark this answers is not in the document]
Comment (nZac) [reply keg_auth/model/__init__.py]: I'm not sure what you mean by multiple api tokens. The tokens are user specific, so a user can only have one token in practice. The optional `token` here is made available mostly for testing purposes.

File: keg_auth/tests/test_model.py [hunk 2]
```diff
@@ -1,17 +1,100 @@
...
+        assert (real_email, token) == (u1.email, u1.token.hash.decode())
+
+    def test_get_user_for_api_token_happy_path(self):
+        u1 = ents.UserWithToken.testing_create(email='test@test.com', token='1234')
+        ents.UserWithToken.testing_create(email='test-2@test.com', token='5678')
+
+        u1_api_token = u1.generate_api_token('1234')
+        assert ents.UserWithToken.get_user_for_api_token(u1_api_token).id == u1.id
+
+    def test_get_user_for_api_token_wrong_token(self):
```
[the remark this answers is not in the document]
Comment (nZac) [reply keg_auth/tests/test_model.py]: That is what this test does. The relevant line is `'{email}.faketoken'` a few lines down, where `faketoken` is the difference.

File: keg_auth_ta/model/entities.py [hunk 3]
```diff
@@ -22,6 +22,10 @@ class UserNoEmail(db.Model, keg_auth.UserMixin, EntityMixin):
     __tablename__ = 'users_no_email'
 
 
+class UserWithToken(db.Model, keg_auth.UserEmailMixin, keg_auth.UserTokenMixin, EntityMixin):
```
[the remark this answers is not in the document]
Comment (nZac) [reply keg_auth_ta/model/entities.py]: Fixed, thank!

File: keg_auth/model/__init__.py [hunk 4]
```diff
@@ -1,3 +1,7 @@
+import binascii
+import base64
+import passlib.hash
```
[the remark this answers is not in the document]
Comment (nZac) [reply keg_auth/model/__init__.py]: Done.

File: keg_auth/tests/test_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tjlevel12) [reply keg_auth/tests/test_model.py]: Ah, you're right, I was misreading that as changing the e-mail address and not the token.

File: keg_auth/model/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjlevel12) [reply keg_auth/model/__init__.py]: Fair enough, I think I was overthinking the goal here.

**Pull request merged**